src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,514 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Represents a data structure in which each element additionally has a priority
+    /// associated with it.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the element.</typeparam>
+    /// <typeparam name="TPriority">The type of the priority.</typeparam>
+    public class PriorityQueue<TElement, TPriority>
+    {
+        /// <summary>
+        /// Represents an implicit heap-ordered complete d-ary tree, stored as an array.
+        /// </summary>
+        private readonly List<(TElement element, TPriority priority)> nodes;
+
+        private const int RootIndex = 0;
+
+        /// <summary>
+        /// Specifies the arity of the d-ary heap, which here is quaternary.
+        /// </summary>
+        private const int Arity = 4;
+
+        /// <summary>
+        /// Creates an empty priority queue.
+        /// </summary>
+        public PriorityQueue()
+        {
+            this.nodes = new List<(TElement, TPriority)>();
+            this.UnorderedItems = new UnorderedItemsCollection(this.nodes);
```

Comment:
This should be created lazily only when the property is accessed the first time, rather than creating it upon construction.